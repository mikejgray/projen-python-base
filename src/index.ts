import { join } from 'path';
import { SampleFile } from 'projen';
import { GitHub, GithubWorkflow } from 'projen/lib/github';
import { JobPermission, JobStep } from 'projen/lib/github/workflows-model';
import { PythonProject, PythonProjectOptions } from 'projen/lib/python';

export const codeCheckout: JobStep = {
  name: 'Checkout',
  id: 'checkout',
  uses: 'actions/checkout@v3',
};

export const precommitCi: JobStep = {
  name: 'Pre-Commit CI',
  id: 'ci',
  run: 'pip install pre-commit\npre-commit run -a',
};

export class PythonPrecommitCI extends GithubWorkflow {
  constructor(github: GitHub) {
    super(github, 'ci');
    this.on({
      pullRequest: {
        types: ['labeled', 'opened', 'synchronize', 'reopened'],
      },
    });

    this.addJob('CI', {
      runsOn: ['ubuntu-latest'],
      name: 'CI',
      steps: [
        codeCheckout,
        precommitCi,
      ],
      permissions: { contents: JobPermission.READ },
      defaults: { run: { shell: 'bash' } },
    });
  }
};

export interface PoetryPythonProjectOptions extends PythonProjectOptions {};

export class PoetryPythonProject extends PythonProject {
  constructor(options: PoetryPythonProjectOptions) {
    let devDeps = ['pylint', 'flake8', 'pydocstyle', 'pycodestyle', 'black', 'mypy', 'bandit'];
    if (options.devDeps) {
      devDeps = [...devDeps, ...options.devDeps];
    }

    super({
      ...options,
      pip: false,
      venv: false,
      setuptools: false,
      poetry: true,
      devDeps: devDeps,
    });

    this.gitignore.addPatterns('.DS_Store');
    new SampleFile(this, '.pre-commit-config.yaml', { sourcePath: join(__dirname, 'files', '.pre-commit-config.yaml') });
    new PythonPrecommitCI(this.github!);

  };
}
