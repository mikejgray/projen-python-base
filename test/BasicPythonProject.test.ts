import { synthSnapshot } from 'projen/lib/util/synth';
import { PoetryPythonProject } from '../src';

function getTestProject(options?: any) {
  return new PoetryPythonProject({
    moduleName: 'test',
    name: 'test',
    authorName: 'Zaphod Beeblebrox',
    authorEmail: 'zaphod@beeblebrox.universe',
    version: '0.0.0',
    ...options,
  });
};

describe('snapshots', () => {
  const project = getTestProject();
  const snap = synthSnapshot(project);

  test('Overall snapshot', () => {
    expect(snap).toMatchSnapshot();
  });

  test('precommit file exists', () => {
    expect(snap['.pre-commit-config.yaml']).toBeDefined();
  });

  test('pyproject.toml exists', () => {
    expect(snap['pyproject.toml']).toBeDefined();
  });

  test('ci workflow exists', () => {
    expect(snap['.github/workflows/ci.yml']).toBeDefined();
  });

  test('dev dependencies are set', () => {
    expect(JSON.stringify(snap['.projen/deps.json'])).toMatch('black');
    expect(JSON.stringify(snap['.projen/deps.json'])).toMatch('pylint');
    expect(JSON.stringify(snap['.projen/deps.json'])).toMatch('flake8');
    expect(JSON.stringify(snap['.projen/deps.json'])).toMatch('pycodestyle');
    expect(JSON.stringify(snap['.projen/deps.json'])).toMatch('pydocstyle');
    expect(JSON.stringify(snap['.projen/deps.json'])).toMatch('mypy');
    expect(JSON.stringify(snap['.projen/deps.json'])).toMatch('bandit');
  });

});
