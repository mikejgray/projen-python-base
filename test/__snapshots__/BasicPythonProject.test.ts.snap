// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots Overall snapshot 1`] = `
Object {
  ".gitattributes": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

/.gitattributes linguist-generated
/.github/workflows/ci.yml linguist-generated
/.github/workflows/pull-request-lint.yml linguist-generated
/.gitignore linguist-generated
/.pre-commit-config.yaml linguist-generated
/.projen/** linguist-generated
/.projen/deps.json linguist-generated
/.projen/files.json linguist-generated
/.projen/tasks.json linguist-generated
/pyproject.toml linguist-generated",
  ".github/workflows/ci.yml": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

name: ci
on:
  pull_request:
    types:
      - labeled
      - opened
      - synchronize
      - reopened
jobs:
  CI:
    name: CI
    runs-on: ubuntu-latest
    permissions:
      contents: read
    defaults:
      run:
        shell: bash
    steps:
      - name: Checkout
        id: checkout
        uses: actions/checkout@v3
      - name: Pre-Commit CI
        id: ci
        run: |-
          pip install pre-commit
          pre-commit run -a
",
  ".github/workflows/pull-request-lint.yml": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

name: pull-request-lint
on:
  pull_request_target:
    types:
      - labeled
      - opened
      - synchronize
      - reopened
      - ready_for_review
      - edited
jobs:
  validate:
    name: Validate PR title
    runs-on: ubuntu-latest
    permissions:
      pull-requests: write
    steps:
      - uses: amannn/action-semantic-pull-request@v5.0.2
        env:
          GITHUB_TOKEN: \${{ secrets.GITHUB_TOKEN }}
        with:
          types: |-
            feat
            fix
            chore
          requireScope: false
",
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
node_modules/
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
!/.github/workflows/pull-request-lint.yml
!/pyproject.toml
/poetry.toml
__pycache__/
*.py[cod]
*$py.class
*.so
.Python
build/
develop-eggs/
dist/
downloads/
eggs/
.eggs/
lib/
lib64/
parts/
sdist/
var/
wheels/
share/python-wheels/
*.egg-info/
.installed.cfg
*.egg
MANIFEST
*.manifest
*.spec
pip-log.txt
pip-delete-this-directory.txt
htmlcov/
.tox/
.nox/
.coverage
.coverage.*
.cache
nosetests.xml
coverage.xml
*.cover
*.py,cover
.hypothesis/
.pytest_cache/
cover/
*.mo
*.pot
*.log
local_settings.py
db.sqlite3
db.sqlite3-journal
instance/
.webassets-cache
.scrapy
docs/_build/
.pybuilder/
target/
.ipynb_checkpoints
profile_default/
ipython_config.py
__pypackages__/
celerybeat-schedule
celerybeat.pid
*.sage.py
.env
.venv
env/
venv/
ENV/
env.bak/
venv.bak/
.spyderproject
.spyproject
.ropeproject
/site
.mypy_cache/
.dmypy.json
dmypy.json
.pyre/
.pytype/

cython_debug/
.DS_Store
!/.pre-commit-config.yaml
!/.github/workflows/ci.yml
",
  ".pre-commit-config.yaml": "# See https://pre-commit.com for more information
# See https://pre-commit.com/hooks.html for more hooks
# Need a rev? Try \`pre-commit autoupdate\` using \`rev: latest\`
repos:
  - repo: https://github.com/pre-commit/pre-commit-hooks
    rev: v4.3.0
    hooks:
      - id: trailing-whitespace
      - id: end-of-file-fixer
      - id: check-added-large-files
      - id: check-ast
  - repo: https://github.com/psf/black
    rev: 22.12.0
    hooks:
      - id: black
        args: [--line-length=119]
  - repo: https://github.com/PyCQA/bandit
    rev: 1.7.4
    hooks:
      - id: bandit
        args: [-ll]
  - repo: https://github.com/PyCQA/flake8
    rev: 6.0.0
    hooks:
      - id: flake8
        args: [--max-line-length=119]
",
  ".projen/deps.json": Object {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
    "dependencies": Array [
      Object {
        "name": "bandit",
        "type": "devenv",
      },
      Object {
        "name": "black",
        "type": "devenv",
      },
      Object {
        "name": "flake8",
        "type": "devenv",
      },
      Object {
        "name": "mypy",
        "type": "devenv",
      },
      Object {
        "name": "projen",
        "type": "devenv",
        "version": "99.99.99",
      },
      Object {
        "name": "pycodestyle",
        "type": "devenv",
      },
      Object {
        "name": "pydocstyle",
        "type": "devenv",
      },
      Object {
        "name": "pylint",
        "type": "devenv",
      },
      Object {
        "name": "pytest",
        "type": "test",
        "version": "6.2.1",
      },
    ],
  },
  ".projen/files.json": Object {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
    "files": Array [
      ".gitattributes",
      ".github/workflows/ci.yml",
      ".github/workflows/pull-request-lint.yml",
      ".gitignore",
      ".pre-commit-config.yaml",
      ".projen/deps.json",
      ".projen/files.json",
      ".projen/tasks.json",
      "poetry.toml",
      "pyproject.toml",
    ],
  },
  ".projen/tasks.json": Object {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
    "env": Object {
      "PATH": "$(echo $(poetry env info -p)/bin:$PATH)",
      "VIRTUAL_ENV": "$(poetry env info -p)",
    },
    "tasks": Object {
      "build": Object {
        "description": "Full release build",
        "name": "build",
        "steps": Array [
          Object {
            "spawn": "default",
          },
          Object {
            "spawn": "pre-compile",
          },
          Object {
            "spawn": "compile",
          },
          Object {
            "spawn": "post-compile",
          },
          Object {
            "spawn": "test",
          },
          Object {
            "spawn": "package",
          },
        ],
      },
      "clobber": Object {
        "condition": "git diff --exit-code > /dev/null",
        "description": "hard resets to HEAD of origin and cleans the local repo",
        "env": Object {
          "BRANCH": "$(git branch --show-current)",
        },
        "name": "clobber",
        "steps": Array [
          Object {
            "exec": "git checkout -b scratch",
            "name": "save current HEAD in \\"scratch\\" branch",
          },
          Object {
            "exec": "git checkout $BRANCH",
          },
          Object {
            "exec": "git fetch origin",
            "name": "fetch latest changes from origin",
          },
          Object {
            "exec": "git reset --hard origin/$BRANCH",
            "name": "hard reset to origin commit",
          },
          Object {
            "exec": "git clean -fdx",
            "name": "clean all untracked files",
          },
          Object {
            "say": "ready to rock! (unpushed commits are under the \\"scratch\\" branch)",
          },
        ],
      },
      "compile": Object {
        "description": "Only compile",
        "name": "compile",
      },
      "default": Object {
        "description": "Synthesize project files",
        "name": "default",
        "steps": Array [
          Object {
            "exec": "python .projenrc.py",
          },
        ],
      },
      "eject": Object {
        "description": "Remove projen from the project",
        "env": Object {
          "PROJEN_EJECTING": "true",
        },
        "name": "eject",
        "steps": Array [
          Object {
            "spawn": "default",
          },
        ],
      },
      "install": Object {
        "description": "Install and upgrade dependencies",
        "name": "install",
        "steps": Array [
          Object {
            "exec": "poetry update",
          },
        ],
      },
      "package": Object {
        "description": "Creates the distribution package",
        "name": "package",
        "steps": Array [
          Object {
            "exec": "poetry build",
          },
        ],
      },
      "post-compile": Object {
        "description": "Runs after successful compilation",
        "name": "post-compile",
      },
      "pre-compile": Object {
        "description": "Prepare the project for compilation",
        "name": "pre-compile",
      },
      "publish": Object {
        "description": "Uploads the package to PyPI.",
        "name": "publish",
        "steps": Array [
          Object {
            "exec": "poetry publish",
          },
        ],
      },
      "publish:test": Object {
        "description": "Uploads the package against a test PyPI endpoint.",
        "name": "publish:test",
        "steps": Array [
          Object {
            "exec": "poetry publish -r testpypi",
          },
        ],
      },
      "test": Object {
        "description": "Run tests",
        "name": "test",
        "steps": Array [
          Object {
            "exec": "pytest",
          },
        ],
      },
    },
  },
  "README.md": "# replace this",
  "poetry.toml": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

[repositories.testpypi]
url = \\"https://test.pypi.org/legacy/\\"
",
  "pyproject.toml": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

[build-system]
requires = [ \\"poetry_core>=1.0.0\\" ]
build-backend = \\"poetry.core.masonry.api\\"

[tool.poetry]
name = \\"test\\"
version = \\"0.0.0\\"
description = \\"\\"
authors = [ \\"Zaphod Beeblebrox <zaphod@beeblebrox.universe>\\" ]
readme = \\"README.md\\"

  [tool.poetry.dependencies]
  python = \\"^3.6\\"

  [tool.poetry.dev-dependencies]
  projen = \\"99.99.99\\"
",
  "test/__init__.py": "__version__ = \\"0.1.0\\"
",
  "test/__main__.py": "from .example import hello

if __name__ == \\"__main__\\":
    name = input(\\"What is your name? \\")
    print(hello(name))
",
  "test/example.py": "def hello(name: str) -> str:
    \\"\\"\\"A simple greeting.
    Args:
        name (str): Name to greet.
    Returns:
        str: greeting message
    \\"\\"\\"
    return f\\"Hello {name}!\\"
",
  "tests/__init__.py": "",
  "tests/test_example.py": "import pytest

from test.example import hello

@pytest.mark.parametrize(
    (\\"name\\", \\"expected\\"),
    [
        (\\"A. Musing\\", \\"Hello A. Musing!\\"),
        (\\"traveler\\", \\"Hello traveler!\\"),
        (\\"projen developer\\", \\"Hello projen developer!\\"),
    ],
)
def test_hello(name, expected):
    \\"\\"\\"Example test with parametrization.\\"\\"\\"
    assert hello(name) == expected
",
}
`;
